### ISSUE TITLE
Dataclass code generation

### ISSUE THREAD

**Issue by lmmx** [lmmx/tubeulator#3]
Following API design (internal handling of request sending) the next task is dataclass generation from JSON component schemas for response handling

The code for this is already written (may have crossed over into comments in issue #2)

- must handle `ref` type (recursion involved?)
- not yet tested on real schema

**Comment by lmmx** [comment]
There're a few mistakes in the generated code:

- Firstly, the call to `validate(o, schema)` uses a `schema` variable that is never defined (but clearly should be set in the `generate_dataclasses` function: [source](https://github.com/lmmx/tubeulator/issues/2#issuecomment-1475278111), [code](https://github.com/lmmx/tubeulator/blob/2786dae5d81c41167957e84c7f7cdc84174c2347/src/tubeulator/codegen.py#L34)).

```py
import json
from dataclasses import dataclass, field
from pathlib import Path
from dataclass_wizard import JSONWizard, LoadMeta
from dataclass_wizard.loaders import fromdict
import jsonschema

@dataclass
class UntitledDeserialiser(JSONWizard):
    isTflService: bool = None
    isFarePaying: bool = None
    isScheduledService: bool = None
    modeName: str = None
    @classmethod
    def from_dict(cls, o):
        jsonschema.validate(o, schema)
        return fromdict(cls, o)
```

**Comment by lmmx** [comment]
I keep taking extended breaks from developing this but I can see that the feature branch has progressed from the master branch here.

For example the first dataclass made from the Line schema now has a name (it previously was UntitledDeserialiser such and such)

```py
import json
from dataclasses import dataclass, field
from pathlib import Path
from dataclass_wizard import JSONWizard, LoadMeta
from dataclass_wizard.loaders import fromdict
import jsonschema

@dataclass
class ModeDeserialiser(JSONWizard):
    """
    Autogenerated from Line::Tfl.Api.Presentation.Entities.Mode
    """
    isTflService: bool = None
    isFarePaying: bool = None
    isScheduledService: bool = None
    modeName: str = None
    
    @classmethod
    def from_dict(cls, o):
        jsonschema.validate(o, schema)
        return fromdict(cls, o)


LoadMeta(raise_on_unknown_json_key=True).bind_to(ModeDeserialiser)
```

I can loop through all the schemas to get a sense of which has the most dataclasses

```sh
for s in $(tubeulator schemas); do echo $s; tubeulator deserialise $s | wc -l; echo; done
```
⇣
```
AccidentStats
125

AirQuality
20

BikePoint
170

crowding
1

Disruptions-Lifts-v2
26

Journey
921

Line
1969

Mode
198

occupancy
321

Place
613

Road
652

Search
228

StopPoint
1603

Vehicle
142
```

**Comment by lmmx** [comment]
The Mode classes show a lot of repetition

```py
class TflApiPresentationEntitiesActiveServiceTypeArrayDeserialiser(JSONWizard):
class ActiveServiceTypeDeserialiser(JSONWizard):
class TflApiPresentationEntitiesActiveServiceTypeArray1Deserialiser(JSONWizard):
class TflApiPresentationEntitiesActiveServiceTypeArray2Deserialiser(JSONWizard):
class TflApiPresentationEntitiesActiveServiceTypeArray3Deserialiser(JSONWizard):
class TflApiPresentationEntitiesPredictionArray4Deserialiser(JSONWizard):
class PredictionDeserialiser(JSONWizard):
class PredictionTimingDeserialiser(JSONWizard):
class TflApiPresentationEntitiesPredictionArray5Deserialiser(JSONWizard):
class TflApiPresentationEntitiesPredictionArray6Deserialiser(JSONWizard):
class TflApiPresentationEntitiesPredictionArray7Deserialiser(JSONWizard):
```

The Vehicle classes show it without the index

```py
class PredictionTimingDeserialiser(JSONWizard):
class PredictionDeserialiser(JSONWizard):
class VehicleMatchDeserialiser(JSONWizard):
class PredictionResponseDeserialiser(JSONWizard):
class PredictionResponseDeserialiser(JSONWizard):
class PredictionResponseDeserialiser(JSONWizard):
class PredictionResponseDeserialiser(JSONWizard):
```

**Comment by lmmx** [comment]
To focus on Vehicle, if we call `tubeulator names` and look at its namespace:

```py
 'Vehicle': {'Tfl': ['Tfl.Api.Presentation.Entities.PredictionTiming'],
             'Tfl-2': ['Tfl.Api.Presentation.Entities.Prediction'],
             'Tfl-3': ['Tfl.Api.Presentation.Entities.VehicleMatch']},
```

we can view the file `src/tubeulator/data/openapi/Vehicle/Vehicle.json` and interpret the classes above:

- `PredictionTimingDeserialiser`
- `PredictionDeserialiser`
- `VehicleMatchDeserialiser`
- `PredictionResponseDeserialiser`

**Comment by lmmx** [comment]
To recap, in the last 2 comments I identified a few aspects of the generator to fix:

- [x] Eliminate repeated classes such as `PredictionResponseDeserialiser` (no issue yet)
- [ ] Ref handling in nested DTOs (#6)

In further edits to `codegen.py` I eliminated all of the duplicate classes.

The full list of classes is below. We 'give up' when we label classes "UNK" (this is good enough I think). The "UNK" prefix is for debugging this algorithm and can be discarded. There are also some classes with long names `_or_` conjoined names, which again I don't think it's important to resolve further. I'm keen to actually **do something** with these classes rather than perfect some precise naming scheme.

In the list below note that there are no longer any groups of similar names, which were finally left in the UNK-prefixed group of names.

For example, in `Journey` there is class `UNK_Get200ApplicationJsonResponseDeserialiser`.

```py
AccidentStats
class AccidentStatsAccidentDetailArrayDeserialiser(JSONWizard):
class AccidentDetailDeserialiser(JSONWizard):
class CasualtyDeserialiser(JSONWizard):
class VehicleDeserialiser(JSONWizard):

AirQuality
class ObjectDeserialiser(JSONWizard):

BikePoint
class PlaceArrayDeserialiser(JSONWizard):
class PlaceDeserialiser(JSONWizard):
class AdditionalPropertiesDeserialiser(JSONWizard):

crowding

Disruptions-Lifts-v2
class UNK_LiftDisruptionDeserialiser(JSONWizard):
class DisruptionsLiftsv2ResponseDeserialiser(JSONWizard):

Journey
class ModeDeserialiser(JSONWizard):
class PathAttributeDeserialiser(JSONWizard):
class InstructionStepDeserialiser(JSONWizard):
class InstructionDeserialiser(JSONWizard):
class ObstacleDeserialiser(JSONWizard):
class PointDeserialiser(JSONWizard):
class PassengerFlowDeserialiser(JSONWizard):
class TrainLoadingDeserialiser(JSONWizard):
class CrowdingDeserialiser(JSONWizard):
class IdentifierDeserialiser(JSONWizard):
class JpElevationDeserialiser(JSONWizard):
class PathDeserialiser(JSONWizard):
class RouteOptionDeserialiser(JSONWizard):
class LineGroupDeserialiser(JSONWizard):
class LineModeGroupDeserialiser(JSONWizard):
class AdditionalPropertiesDeserialiser(JSONWizard):
class PlaceDeserialiser(JSONWizard):
class StopPointDeserialiser(JSONWizard):
class RouteSectionNaptanEntrySequenceDeserialiser(JSONWizard):
class RouteSectionDeserialiser(JSONWizard):
class DisruptionDeserialiser(JSONWizard):
class PlannedWorkDeserialiser(JSONWizard):
class LegDeserialiser(JSONWizard):
class FareTapDetailsDeserialiser(JSONWizard):
class FareTapDeserialiser(JSONWizard):
class FareDeserialiser(JSONWizard):
class FareCaveatDeserialiser(JSONWizard):
class JourneyFareDeserialiser(JSONWizard):
class JourneyDeserialiser(JSONWizard):
class ValidityPeriodDeserialiser(JSONWizard):
class LineStatusDeserialiser(JSONWizard):
class MatchedRouteDeserialiser(JSONWizard):
class LineServiceTypeInfoDeserialiser(JSONWizard):
class LineDeserialiser(JSONWizard):
class JourneyPlannerCycleHireDockingStationDataDeserialiser(JSONWizard):
class TimeAdjustmentDeserialiser(JSONWizard):
class TimeAdjustmentsDeserialiser(JSONWizard):
class SearchCriteriaDeserialiser(JSONWizard):
class JourneyVectorDeserialiser(JSONWizard):
class ItineraryResultDeserialiser(JSONWizard):
class MetaModesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class UNK_Get200ApplicationJsonResponseDeserialiser(JSONWizard):

Line
class ModeDeserialiser(JSONWizard):
class StatusSeverityDeserialiser(JSONWizard):
class PassengerFlowDeserialiser(JSONWizard):
class TrainLoadingDeserialiser(JSONWizard):
class CrowdingDeserialiser(JSONWizard):
class IdentifierDeserialiser(JSONWizard):
class LineGroupDeserialiser(JSONWizard):
class LineModeGroupDeserialiser(JSONWizard):
class AdditionalPropertiesDeserialiser(JSONWizard):
class PlaceDeserialiser(JSONWizard):
class StopPointDeserialiser(JSONWizard):
class RouteSectionNaptanEntrySequenceDeserialiser(JSONWizard):
class RouteSectionDeserialiser(JSONWizard):
class DisruptionDeserialiser(JSONWizard):
class ValidityPeriodDeserialiser(JSONWizard):
class LineStatusDeserialiser(JSONWizard):
class MatchedRouteDeserialiser(JSONWizard):
class LineServiceTypeInfoDeserialiser(JSONWizard):
class LineDeserialiser(JSONWizard):
class MatchedStopDeserialiser(JSONWizard):
class StopPointSequenceDeserialiser(JSONWizard):
class OrderedRouteDeserialiser(JSONWizard):
class RouteSequenceDeserialiser(JSONWizard):
class LineRouteSectionDeserialiser(JSONWizard):
class MatchedRouteSectionsDeserialiser(JSONWizard):
class RouteSearchMatchDeserialiser(JSONWizard):
class RouteSearchResponseDeserialiser(JSONWizard):
class IntervalDeserialiser(JSONWizard):
class StationIntervalDeserialiser(JSONWizard):
class KnownJourneyDeserialiser(JSONWizard):
class TwentyFourHourClockTimeDeserialiser(JSONWizard):
class ServiceFrequencyDeserialiser(JSONWizard):
class PeriodDeserialiser(JSONWizard):
class ScheduleDeserialiser(JSONWizard):
class TimetableRouteDeserialiser(JSONWizard):
class TimetableDeserialiser(JSONWizard):
class DisambiguationOptionDeserialiser(JSONWizard):
class DisambiguationDeserialiser(JSONWizard):
class TimetableResponseDeserialiser(JSONWizard):
class PredictionTimingDeserialiser(JSONWizard):
class PredictionDeserialiser(JSONWizard):
class MetaModesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class UNK_Get200ApplicationJsonResponseDeserialiser(JSONWizard):
class MetaSeverityGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class MetaDisruptionCategoriesDeserialiser(JSONWizard):
class MetaServiceTypesDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_ModemodesGet200_or_RouteGet200_or_idsRouteGet200_or_ModemodesRouteGet200_or_idsStatusstartDateToendDateGet200_or_idsStatusGet200_or_StatusseverityGet200_or_ModemodesStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idStopPointsGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsDisruptionGet200_or_ModemodesDisruptionGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsDisruptionGet200_or_ModemodesDisruptionGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsArrivalsstopPointIdGet200_or_idsArrivalsGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsArrivalsstopPointIdGet200_or_idsArrivalsGet200ApplicationJsonResponseDeserialiser(JSONWizard):

Mode
class ActiveServiceTypeArrayDeserialiser(JSONWizard):
class ActiveServiceTypeDeserialiser(JSONWizard):
class PredictionArrayDeserialiser(JSONWizard):
class PredictionDeserialiser(JSONWizard):
class PredictionTimingDeserialiser(JSONWizard):

occupancy
class BayDeserialiser(JSONWizard):
class CarParkOccupancyDeserialiser(JSONWizard):
class ChargeConnectorOccupancyDeserialiser(JSONWizard):
class BikePointOccupancyDeserialiser(JSONWizard):
class CarParkGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class ChargeConnectoridsGet200_or_ChargeConnectorGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class ChargeConnectoridsGet200_or_ChargeConnectorGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class BikePointsidsGet200ApplicationJsonResponseDeserialiser(JSONWizard):

Place
class PlaceCategoryDeserialiser(JSONWizard):
class AdditionalPropertiesDeserialiser(JSONWizard):
class PlaceDeserialiser(JSONWizard):
class PassengerFlowDeserialiser(JSONWizard):
class TrainLoadingDeserialiser(JSONWizard):
class CrowdingDeserialiser(JSONWizard):
class IdentifierDeserialiser(JSONWizard):
class LineGroupDeserialiser(JSONWizard):
class LineModeGroupDeserialiser(JSONWizard):
class StopPointDeserialiser(JSONWizard):
class ObjectDeserialiser(JSONWizard):
class MetaCategoriesGet200_or_MetaPlaceTypesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class UNK_Get200ApplicationJsonResponseDeserialiser(JSONWizard):
class MetaCategoriesGet200_or_MetaPlaceTypesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class TypetypesGet200_or_idGet200_or_SearchGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class TypetypesGet200_or_idGet200_or_SearchGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class ApplicationJsonResponseDeserialiser(JSONWizard):
class TypetypesGet200_or_idGet200_or_SearchGet200ApplicationJsonResponseDeserialiser(JSONWizard):

Road
class RoadCorridorDeserialiser(JSONWizard):
class DbGeographyWellKnownValueDeserialiser(JSONWizard):
class DbGeographyDeserialiser(JSONWizard):
class StreetSegmentDeserialiser(JSONWizard):
class StreetDeserialiser(JSONWizard):
class RoadProjectDeserialiser(JSONWizard):
class RoadDisruptionLineDeserialiser(JSONWizard):
class RoadDisruptionImpactAreaDeserialiser(JSONWizard):
class RoadDisruptionScheduleDeserialiser(JSONWizard):
class RoadDisruptionDeserialiser(JSONWizard):
class ObjectDeserialiser(JSONWizard):
class StatusSeverityDeserialiser(JSONWizard):
class Get200_or_idsGet200_or_idsStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class Get200_or_idsGet200_or_idsStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class Get200_or_idsGet200_or_idsStatusGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsDisruptionGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class MetaCategoriesDeserialiser(JSONWizard):
class MetaSeveritiesGet200ApplicationJsonResponseDeserialiser(JSONWizard):

Search
class SearchMatchDeserialiser(JSONWizard):
class SearchResponseDeserialiser(JSONWizard):
class MetaSearchProvidersDeserialiser(JSONWizard):
class MetaCategoriesDeserialiser(JSONWizard):
class MetaSortsDeserialiser(JSONWizard):

StopPoint
class PlaceCategoryDeserialiser(JSONWizard):
class ModeDeserialiser(JSONWizard):
class PassengerFlowDeserialiser(JSONWizard):
class TrainLoadingDeserialiser(JSONWizard):
class CrowdingDeserialiser(JSONWizard):
class IdentifierDeserialiser(JSONWizard):
class LineGroupDeserialiser(JSONWizard):
class LineModeGroupDeserialiser(JSONWizard):
class AdditionalPropertiesDeserialiser(JSONWizard):
class PlaceDeserialiser(JSONWizard):
class StopPointDeserialiser(JSONWizard):
class LineServiceTypeInfoDeserialiser(JSONWizard):
class LineSpecificServiceTypeDeserialiser(JSONWizard):
class LineServiceTypeDeserialiser(JSONWizard):
class PredictionTimingDeserialiser(JSONWizard):
class PredictionDeserialiser(JSONWizard):
class ArrivalDepartureDeserialiser(JSONWizard):
class StopPointRouteSectionDeserialiser(JSONWizard):
class DisruptedPointDeserialiser(JSONWizard):
class StopPointsResponseDeserialiser(JSONWizard):
class SearchMatchDeserialiser(JSONWizard):
class SearchResponseDeserialiser(JSONWizard):
class ObjectDeserialiser(JSONWizard):
class UNK_Get200ApplicationJsonResponseDeserialiser(JSONWizard):
class MetaCategoriesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class MetaStopTypesDeserialiser(JSONWizard):
class MetaModesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_idCrowdinglineGet200_or_TypetypesGet200_or_TypetypesPagepageGet200_or_idCanReachOnLinelineIdGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idPlaceTypesGet200_or_stopPointIdTaxiRanksGet200_or_stopPointIdCarParksGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_idCrowdinglineGet200_or_TypetypesGet200_or_TypetypesPagepageGet200_or_idCanReachOnLinelineIdGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_idCrowdinglineGet200_or_TypetypesGet200_or_TypetypesPagepageGet200_or_idCanReachOnLinelineIdGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_idCrowdinglineGet200_or_TypetypesGet200_or_TypetypesPagepageGet200_or_idCanReachOnLinelineIdGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class ServiceTypesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idArrivalsGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idArrivalDeparturesGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idsGet200_or_idCrowdinglineGet200_or_TypetypesGet200_or_TypetypesPagepageGet200_or_idCanReachOnLinelineIdGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idRouteGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class ModemodesDisruptionGet200_or_idsDisruptionGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class ModemodesDisruptionGet200_or_idsDisruptionGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class UNK_idDirectionTotoStopPointIdGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idPlaceTypesGet200_or_stopPointIdTaxiRanksGet200_or_stopPointIdCarParksGet200ApplicationJsonResponseDeserialiser(JSONWizard):
class idPlaceTypesGet200_or_stopPointIdTaxiRanksGet200_or_stopPointIdCarParksGet200ApplicationJsonResponseDeserialiser(JSONWizard):

Vehicle
class PredictionTimingDeserialiser(JSONWizard):
class PredictionDeserialiser(JSONWizard):
class VehicleMatchDeserialiser(JSONWizard):
class idsArrivalsGet200ApplicationJsonResponseDeserialiser(JSONWizard):
```

**Comment by lmmx** [comment]
I regenerated all the dataclasses with

```sh
(tubeulator) louis 🚶 ~/dev/tubeulator/src/tubeulator/generated $
for s in $(tubeulator schemas); do echo $s; tubeulator deserialise $s > "$s.py"; done
```

**Comment by lmmx** [comment]
Once you run `black` on the files, the sole flake8 complaints are:

```
 F401 'dataclasses.field' imported but unused
 F401 'json' imported but unused
 F401 'pathlib.Path' imported but unused
 F811 redefinition of unused 'Path' from line 3
 F821 undefined name 'Any'
 F821 undefined name 'schema'
 W391 blank line at end of file
```

**Comment by lmmx** [comment]
Merged #5
